types/externs.d.ts

```diff
@@ -76,6 +76,25 @@ declare global {
       T[P];
   };
 
+  /** Recursively makes all properties of T read-only. */
+  export type Immutable<T> =
+    T extends Function ? T :
+    T extends ReadonlyArray<infer R> ? ImmutableArray<R> :
+    T extends Map<infer K, infer V> ? ImmutableMap<K, V> :
+    T extends Set<infer M> ? ImmutableSet<M> :
+    T extends object ? ImmutableObject<T> :
+    T
+
+  // Intermediate immutable interfaces. Prefer e.g. Immutable<Set<T>> over direct use.
+  // TODO: switch to recursive type references when using tsc ≥ 3.7
+  // see https://github.com/microsoft/TypeScript/issues/13923#issuecomment-557509399
+  interface ImmutableArray<T> extends ReadonlyArray<Immutable<T>> {}
```

Comment:
when we upgrade tsc, this can become 
`type ImmutableArray<T> = ReadonlyArray<Immutable<T>>;`, which means things like the built-in `ReadonlyArray<string>` will be transparently switchable with `Immutable<Array<string>>`, giving a lot better ergonomics and functionality